examples/constrained/binary_kernel_svm_with_intercept.py

```diff
@@ -19,72 +19,193 @@
+    Time : :math:`\mathbb{O}(nk)`.
+
+  Significant speedup can be observed on parallel architectures
+  compared to this worst case complexity.
+  """
+
+  def init_params(self, X, Y=None):
+    """Return params from data matrices X and Y.
+    
+    Args:
+      X: array of shape (num_samples_X, num_features).
+      Y: (optional) array of shape (num_samples_Y, num_features).
+         Default to X if not given.
+    """
+    if Y is None:
+      return X, X
+    return X, Y
+
+  def matvec(self, params, vec):
+    X, Y = params
+    return jnp.dot(X, jnp.dot(Y.T, vec))
+
+  def get_dense(self, params):
+    X, Y = params
+    return X @ Y.T
+
+
+def binary_kernel_svm_skl(X, y, C, kernel):
+  print(f"Solve SVM with sklearn.svm.SVC: ")
+  kernel_params = kernel.init_params(X)
+  K = kernel.get_dense(kernel_params)
+  svc = svm.SVC(kernel="precomputed", C=C, tol=FLAGS.tol).fit(K, y)
+  dual_coef = onp.zeros(K.shape[0])
+  dual_coef[svc.support_] = svc.dual_coef_[0]
+  return dual_coef
+
+
+def binary_kernel_svm_pg(X, y, C, kernel):
+  print(f"Solve SVM with Projected Gradient: ")
+
+  def objective_fun(beta, kernel_params, y):
+    """Dual objective of binary kernel SVMs with intercept."""
+    # The dual objective is:
+    # fun(beta) = 0.5 beta^T K beta - beta^T y
+    # subject to
+    # sum(beta) = 0
+    # 0 <= beta_i <= C if y_i = 1
+    # -C <= beta_i <= 0 if y_i = -1
+    # where C = 1.0 / lam
+    # and K = X X^T
+    Kbeta = kernel.matvec(kernel_params, beta)
+    return 0.5 * jnp.dot(beta, Kbeta) - jnp.dot(beta, y)
+
+  # Define projection operator.
+  w = jnp.ones(X.shape[0])
+
+  def proj(beta, C):
+    box_lower = jnp.where(y == 1, 0, -C)
+    box_upper = jnp.where(y == 1, C, 0)
+    proj_params = (box_lower, box_upper, w, 0.0)
+    return projection.projection_box_section(beta, proj_params)
+
+  kernel_params = kernel.init_params(X)
+
+  # Run solver.
+  beta_init = jnp.ones(X.shape[0])
+  solver = ProjectedGradient(fun=objective_fun,
+                             projection=proj,
+                             tol=FLAGS.tol, maxiter=500, verbose=FLAGS.verbose)
+  beta_fit = solver.run(beta_init, hyperparams_proj=C, kernel_params=kernel_params, y=y).params
+
+  return beta_fit
+
+
+def binary_kernel_svm_osqp(X, y, C, kernel):
   # The dual objective is:
   # fun(beta) = 0.5 beta^T K beta - beta^T y
   # subject to
   # sum(beta) = 0
   # 0 <= beta_i <= C if y_i = 1
   # -C <= beta_i <= 0 if y_i = -1
   # where C = 1.0 / lam
-  return 0.5 * jnp.dot(beta, jnp.dot(K, beta)) - jnp.dot(beta, y)
 
+  print(f"Solve SVM with OSQP: ")
 
-def binary_kernel_svm_skl(K, y, lam, tol=1e-5):
-  svc = svm.SVC(kernel="precomputed", C=1.0 / lam).fit(K, y)
-  dual_coef = onp.zeros(K.shape[0])
-  dual_coef[svc.support_] = svc.dual_coef_[0]
-  return dual_coef
+  kernel_params = kernel.init_params(X)
+
+  def matvec_Q(kernel_params, beta):
+    return kernel.matvec(kernel_params, beta)
+  
+  def matvec_A(_, beta):
+    return beta, jnp.sum(beta)
```

Comment:
Why do you return a tuple here?

Reply:
They are two types of constraints:

```python
0 <= y*beta <= C
sum_i beta_i = 0
```

So it feels natural to return a tuple: the first element corresponds to identity matrix (in matvec form) and the other to the vector full of ones (in mnatvec form once again).

The shapes of the output of `matvec_A` and of `l,u` must agree. Since the algorithm supports pytrees (which include tuples) transparently I figured it was maybe more clear to use them. We could have returned vector by concatenating data but it's uglier IMO.

Reply:
Ok thanks for the clarification. It's just a matter of writing a comment explaining how this works then.